### PR TITLE
Update cmake to indicate which libraries are optional, and not fail when they cannot be configured

### LINKED ISSUE #212: Build issue: SUNDIALS without KLU
### Build error message(s)

```console
 CMake Error at cmake/GridkitAddLibrary.cmake:57 (target_link_libraries):
   Target "solvers_dyn_obj_shared" links to:

     SUNDIALS::sunlinsolklu

   but the target was not found.  Possible reasons include:

     * There is a typo in the target name.
     * A find_package call is missing for an IMPORTED target.
     * An ALIAS target is missing.

 Call Stack (most recent call first):
   src/Solver/Dynamic/CMakeLists.txt:8 (gridkit_add_library)

 CMake Error at cmake/GridkitAddLibrary.cmake:69 (target_link_libraries):
   Target "solvers_dyn_shared" links to:

     SUNDIALS::sunlinsolklu

   but the target was not found.  Possible reasons include:

     * There is a typo in the target name.
     * A find_package call is missing for an IMPORTED target.
     * An ALIAS target is missing.

 Call Stack (most recent call first):
   src/Solver/Dynamic/CMakeLists.txt:8 (gridkit_add_library)

 Generating done
```

### GridKit™ version

0.1.0

### System and environment details

When building with `GRIDKIT_ENABLE_SUNDIALS=ON` and `GRIDKIT_ENABLE_SUNDIALS_SPARSE=OFF`.

### Additional information

We'll need to either require that SUNDIALS is built with KLU or add more checks in the CMake and cpp files to handle the case when it isn't available.

### PR DESCRIPTION
**Pull request by alexander-novo**
## Description
 
Fixes #212.

There's currently an issue with how optional library components and executables are setup in the cmake. These libraries and executables are guarded by if statements checking if their optional components are installed before including their folders. If a library changes which optional components it needs, then all of the if statements guarding its folder and all derivative executables need to change. This is also fragile, since it requires all optional components to be identified ahead of time (see #212 where it was assumed `sunlinsolklu` wasn't optional).
 
 ## Proposed changes
 
Now, the `gridkit_add_library` macro allows an `OPTIONAL` parameter and will check the existence of all linked libraries. If the linked libraries exist - nothing different happens, but if a library which must be linked doesn't exist and the library is marked as `OPTIONAL`, then that target will simply be ignored.

There is a new `gridkit_add_executable` macro as well which does a similar process - if an executable depends on a disabled optional library, then the executable target will just become ignored.
 
 ## Checklist
 
- [x] All tests pass.
- [x] Code compiles cleanly with flags `-Wall -Wpedantic -Wconversion -Wextra`.
- [x] The new code follows GridKit™ style guidelines.
- [ ] There are unit tests for the new code.
- [ ] The new code is documented.
- [x] The feature branch is rebased with respect to the target branch.
 
 ## Further comments
 
Only the `PowerElectronics` examples have been configured using the new `gridkit_add_executable` macro, since these were the ones giving me trouble - due to a cmake file looking for the `KLU` target instead of `sunlinsolklu` (which was actually what was needed).

### CODE REVIEW

**Comment by nkoukpaizan** [comment]
Tests pass in CI and on Frontier. I need to take a closer look for edge cases. 

This does not entirely fix #212. It partially addresses the CMake component, but the other part of the issue is that [Ida.cpp](https://github.com/ORNL/GridKit/blob/develop/src/Solver/Dynamic/Ida.cpp) has code for both the dense and the sparse interface. So, it breaks when SUNDIALS is not built with KLU. Marking it as optional altogether is not enough, because there are other dependencies that would also have to be marked optional. I'd say this is out of scope for this PR.
``` 
     22    -- Disabling optional GridKit library solvers_dyn due to missing library: SUNDIALS::sunlinsolklu
  >> 23    CMake Error at cmake/GridkitAddLibrary.cmake:28 (message):
     24      Failed to configure required GridKit library solvers_opt due to missing
     25      library: GRIDKIT::solvers_dyn
     26    Call Stack (most recent call first):
     27      src/Solver/Optimization/CMakeLists.txt:8 (gridkit_add_library)
```

**Comment by alexander-novo** [comment]
That's true, there is a bunch of code in `Ida.cpp` that needs to be revamped. There is also a dependence on IDAS that could potentially be removed as well.

That issue should probably be renamed, because right now it's labeled as a "build issue" and as of right now it's impossible to build `Ida.cpp` without `sunlinsolklu` and without changing the code of `Ida.cpp`. This PR at least fixes cmake attempting to build `Ida.cpp` without `sunlinsolklu`

**Comment by alexander-novo** [comment]
I also wanted to add an example of what it looks like when an optional library is missing its dependency:

```
root@907866574cc8:/workspaces/GridKit/build# cmake .. -DGRIDKIT_ENABLE_SUNDIALS=OFF
-- Disabling optional GridKit library solvers_dyn due to missing library: SUNDIALS::nvecserial
-- Disabling GridKit executable rlcircuit due to disabled optional library: GRIDKIT::solvers_dyn
-- Disabling GridKit executable microgrid due to disabled optional library: GRIDKIT::solvers_dyn
-- Disabling GridKit executable scalemicrogrid due to disabled optional library: GRIDKIT::solvers_dyn
-- Configuring done (1.1s)
```

**Comment by Steven-Roberts** [comment]
> That's true, there is a bunch of code in `Ida.cpp` that needs to be revamped.

I'm curious what you have in mind for this. Perhaps you could make an issue out of this.

File: cmake/GridkitAddExecutable.cmake [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+macro(gridkit_add_executable target)
+
+  set(options TEST)
+  set(multiValueArgs SOURCES LINK_LIBRARIES COMPILE_OPTIONS)
+
+  # parse arguments
+  cmake_parse_arguments(gridkit_add_executable
+    "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  # Before going any further and configuring this library - make sure we can actually link its requirements
```
Comment (shakedregev) [inline cmake/GridkitAddExecutable.cmake]:
```suggestion
  # Before going any further and configuring this library, make sure we can actually link its requirements.
```

File: cmake/GridkitAddLibrary.cmake [hunk 2]
```diff
@@ -8,14 +8,29 @@
...
-  set(options STATIC_ONLY SHARED_ONLY)
+  set(options STATIC_ONLY SHARED_ONLY OPTIONAL)
   set(oneValueArgs OUTPUT_NAME)
   set(multiValueArgs SOURCES LINK_LIBRARIES INCLUDE_DIRECTORIES COMPILE_OPTIONS)
 
   # parse arguments
   cmake_parse_arguments(gridkit_add_library
     "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
 
+  # Before going any further and configuring this library - make sure we can actually link its requirements
```
Comment (shakedregev) [inline cmake/GridkitAddLibrary.cmake]:
```suggestion
  # Before going any further and configuring this library, make sure we can actually link its requirements.
```

**Review by shakedregev — COMMENTED** [review]
Still need to test. Minor comments.

**Review by shakedregev — APPROVED** [review]
Tests pass.

**Comment by alexander-novo** [comment]
@pelesh mentioned on slack:

> All external dependencies are optional.

So it might be good to go through after this PR and update all GridKit libraries which depend on external dependencies as optional.

**Review by pelesh — CHANGES_REQUESTED** [review]
The scope of this PR is not quite well defined. Having a macro `gridkit_add_executable` is certainly helpful, but how does it fix #212? I suggest breaking this PR into specific focused PRs addressing well-defined issues.

**Comment by pelesh** [comment]
Closing in favor of #218. I suggest we document and ideas from this PR is corresponding issues and submit more focused PRs once we decide how to proceed.
